follower/execution/follower.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package execution
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage/badger/operation"
+
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+type blockReader interface {
+	Read(blockID flow.Identifier) ([]byte, error)
+}
+
+type unmarshaler interface {
+	Unmarshal(compressed []byte, value interface{}) error
+}
+
+type Follower struct {
+	log zerolog.Logger
+
+	blocks blockReader
+	codec  unmarshaler
+	db     *badger.DB
+
+	block BlockData
+}
+
+func New(log zerolog.Logger, blocks blockReader, codec unmarshaler, db *badger.DB) *Follower {
+	f := Follower{
+		log: log,
+
+		blocks: blocks,
+		codec:  codec,
+		db:     db,
+	}
+
+	return &f
+}
+
+func (f *Follower) Height() uint64 {
+	if f.block.Block != nil {
+		return f.block.Block.Header.Height
+	}
+	return math.MaxUint64
+}
+
+func (f *Follower) Update() (*ledger.TrieUpdate, error) {
+	if len(f.block.TrieUpdates) == 0 {
+		// No more trie updates are available.
+		return nil, dps.ErrTimeout
+	}
+
+	// Copy next update to be returned.
+	update := f.block.TrieUpdates[0]
+
+	// Move the slice forward by popping the first element.
+	f.block.TrieUpdates = f.block.TrieUpdates[1:]
+
+	return update, nil
+}
+
+// Block returns the latest sealed block data.
+func (f *Follower) Block() BlockData {
+	return f.block
+}
+
+func (f *Follower) OnBlockFinalized(finalizedBlockID flow.Identifier) {
+	for {
+		b, err := f.blocks.Read(finalizedBlockID)
+		if err != nil {
+			// The block data is not yet available. Retry until it becomes available.
+			time.Sleep(1 * time.Second)
```

Comment:
This is definitely not where we should have the retry logic. Everything should be driven by the business logic, not this component.

Reply:
IMO those are two separate things. This sleep is there to wait until the block that has been finalized is available in the bucket, if it isn't already.

Meanwhile, if the mapper attempts to call the Follower Chain's methods, since those try to read from the database only once the Execution Follower's height is at the requested height, the Chain will simply return timeout errors as expected.

The logic of fetching files from Google Cloud should be completely separate from the mapper's FSM, in my opinion 🤔

Reply:
There is no correlation between when `OnBlockFinalized` is called, and when the block is executed. Some blocks need long, some blocks are fast. There is also no need to write anything to the disk, that will slow down indexing significantly.

The execution follower should just poll regularly and load all blocks that are available and keep them in memory in a buffer. Regardless of whether they are already finalized or not. When a block is fully consumed (all trie updates have been read), it should be removed from the buffer and it should try to add the next block to the buffer.

I think a condition variable / semaphore, with a ticker for polling, is the appropriate mechanism here, not an idle sleep.

Reply:
Makes sense to me, I'll implement that change as well then.